autopush/utils.py

```diff
@@ -198,3 +223,272 @@ def parse_user_agent(agent_string):
+    value so that an application server may delete the message. It's used as
+    part of the non-versioned sort-key. Due to this, its an encrypted value
+    that contains the necessary information to derive the location of this
+    precise message in the appropriate message table.
+
+    """
+    uaid = attrib()  # type: uuid.UUID
+    channel_id = attrib()  # type: uuid.UUID
+    ttl = attrib()  # type: int
+    data = attrib(default=None)
+    headers = attrib(default=None)  # type: dict
+    timestamp = attrib(default=Factory(lambda: int(time.time())))
+    topic = attrib(default=None)
+
+    message_id = attrib(default=None)  # type: str
+
+    # Not an alias for message_id, for backwards compat and cases where an old
+    # message with any update_id should be removed.
+    update_id = attrib(default=None)  # type: str
+
+    def generate_message_id(self, fernet):
+        """Generate a message-id suitable for accessing the message
+
+        For non-topic messages, no sort_key version is currently used and the
+        message-id is:
+
+            Encrypted(m : uaid.hex : channel_id.hex)
+
+        For topic messages, a sort_key version of 01 is used, and the topic
+        is included for reference:
+
+            Encrypted(01 : uaid.hex : channel_id.hex : topic)
+
+        This is a blocking call.
+
+        :type fernet: cryptography.fernet.Fernet
+
+        """
+        if self.topic:
+            msg_key = ":".join(["01", self.uaid.hex, self.channel_id.hex,
+                                self.topic])
+        else:
+            msg_key = ":".join(["m", self.uaid.hex, self.channel_id.hex])
+        self.message_id = fernet.encrypt(msg_key.encode('utf8'))
+        self.update_id = self.message_id
+        return self.message_id
+
+    @staticmethod
+    def parse_decrypted_message_id(decrypted_token):
+        """Parses a decrypted message-id into component parts
+
+        :type decrypted_token: str
+        :rtype: dict
+
+        """
+        topic = None
+        if decrypted_token.startswith("01:"):
+            info = decrypted_token.split(":")
+            if len(info) != 4:
+                raise InvalidTokenException("Incorrect number of token parts.")
+            api_ver, uaid, chid, topic = info
+        else:
+            info = decrypted_token.split(":")
+            if len(info) != 3:
+                raise InvalidTokenException("Incorrect number of token parts.")
+            kind, uaid, chid = decrypted_token.split(":")
+            if kind != "m":
+                raise InvalidTokenException("Incorrect token kind.")
+        return dict(
+            uaid=uaid,
+            chid=chid,
+            topic=topic,
+        )
+
+    def cleanup_headers(self):
+        """Sanitize the headers for this notification
+
+        This only needs to be run when creating a notification from passed
+        in application server headers.
+
+        """
+        self.headers = sanitize_webpush_headers(self.headers)
```

Comment:
We seem to only be using this function to call `sanitize_webpush_headers`. Can we collapse a function?